presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java

```diff
@@ -46,22 +49,57 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Pla
         }
 
         // Assign 'statsEquivalentPlanNode' to plan nodes
-        plan = SimplePlanRewriter.rewriteWith(new Rewriter(), plan);
+        plan = SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator), plan, new Context());
 
         // TODO: Fetch and cache history based statistics of all plan nodes, so no serial network calls happen later.
         return plan;
     }
 
     private static class Rewriter
-            extends SimplePlanRewriter<Void>
+            extends SimplePlanRewriter<Context>
     {
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
         @Override
-        public PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
+        public PlanNode visitPlan(PlanNode node, RewriteContext<Context> context)
         {
             if (!node.getStatsEquivalentPlanNode().isPresent()) {
-                node = node.assignStatsEquivalentPlanNode(Optional.of(node));
+                PlanNode nodeWithLimit = node;
+                // A LIMIT node can affect statistics of several children nodes, so let's put the limit node in the stats equivalent plan node.
+                // For example:
+                // `SELECT * FROM nation where <filter> LIMIT 5` vs `SELECT * FROM nation where <filter>`
+                // Filter nodes in both examples will have different output stats because once LIMIT node gets 5 rows, it will stop all operators, and
+                // their statistics will be "incomplete".
+                if (!context.get().getLimits().empty()) {
```

Comment:
1. This will mark more than necessary because blocking operators will not really be affected by the limit (e.g. hash builder, global aggregation)
2. It's possible to have multiple limits affecting a node, and the nearest one isn't necessarily the one that will short-circuit the query e.g. `SELECT * from  (SELECT * FROM A LIMIT 1000000) join b limit 2`. it's possible that out of 1000000 entries from table a, there will only be one matching a row from b, so the limit 2 will never do anything.  But it's also possible that all will match a row from b, but the computation will be short circuited once 2 matches are found for the output.

Reply:
Discussed offline